### PR TITLE
[Merged by Bors] - feat: port field_simp tactic

### PR DESCRIPTION
**Pull request by dwrensha**
* Ports the `field_simp` tactic and tests from mathlib3.
* Adds a new test taken from minif2f (but with rats instead of reals, for simplicity): https://github.com/openai/miniF2F/blob/4e433ff5cadff23f9911a2bb5bbab2d351ce5554/lean/src/valid.lean#L524-L530.
* Uses `field_simp` in two already-ported files, as directed by port notes.
* Adds an explicit priority to a few `field_simps` lemmas. This is necessary to make all of the examples work.

### CODE REVIEW

File: Mathlib/Tactic/FieldSimp.lean [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+initialize registerTraceClass `Tactic.field_simp
+
+/-- Constructs a trace message for the `discharge` function. -/
+private def dischargerTraceMessage (prop: Expr) : Except ε (Option Expr) → SimpM MessageData
+| .error _ | .ok none => return m!"{crossEmoji} discharge {prop}"
+| .ok (some _) => return m!"{checkEmoji} discharge {prop}"
+
+/-- Discharge strategy for the field_simp tactic. -/
+partial def discharge (prop : Expr) : SimpM (Option Expr) :=
```
Comment (dwrensha) [inline Mathlib/Tactic/FieldSimp.lean:30]: I used the default discharger implementation as a model for this function. https://github.com/leanprover/lean4/blob/1f41b912064e49979bfefa5874ea53cb25934a06/src/Lean/Meta/Tactic/Simp/Main.lean#L815-L835

File: Mathlib/Tactic/FieldSimp.lean [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+  let ctx : Simp.Context := {
+     simpTheorems := #[thms, thms0]
+     congrTheorems := (← getSimpCongrTheorems)
+     config := cfg
+  }
+  let r ← elabSimpArgs (sa.getD ⟨.missing⟩) ctx (eraseLocal := false) .simp
+
+  let _ ← simpLocation r.ctx dis loc
+  pure ()
```
Comment (fpvandoorn) [inline Mathlib/Tactic/FieldSimp.lean:155]:
```suggestion
```

**Review by fpvandoorn — COMMENTED** [review]
Since there seem to be quite some differences between the mathlib3 and mathlib4 implementations of `field_simps`, it might be good to test this on more examples, if it is reasonably easy to add more examples...

I saw three commented out cases in `test/ring`. If I add 
```
import Mathlib.Tactic.FieldSimp
import Mathlib.Data.Real.Basic
```
the first two work, but are incredibly slow, and I think the third one is supposed to further simplify `2 * x ^ 3 * 2 / (24 * x)` and fails.

**Comment by dwrensha** [comment]
Re-added a call to `NormNum.derive` in `discharge`. That brings the implementation more in line with mathlib3, and gets the previously-broken real-numbers example in test/ring.lean to work.
(Note that #1711 the real numbers get mocked out in that test file. My fix works on both the real real numbers and on the mocked-out one.)

File: Mathlib/Tactic/FieldSimp.lean [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+/-- Discharge strategy for the field_simp tactic. -/
+partial def discharge (prop : Expr) : SimpM (Option Expr) :=
+  withTraceNode `Tactic.field_simp (dischargerTraceMessage prop) do
+    if let some r ← Simp.dischargeUsingAssumption? prop then
+      return some r
+
+    let prop : Q(Prop) ← (do pure prop)
+    let pf? ← match prop with
+    | ~q(($e : $α) ≠ $b) =>
```
Comment (dwrensha) [inline Mathlib/Tactic/FieldSimp.lean:40]: The mathlib3 version instead matches on `$e ≠ 0`. I couldn't figure out a way to make that work with `Qq`. (I could get it to typecheck here, but then it failed to actually match the expressions that we need it to. This more general match should be fine. The cost is that we might call `NormNum.derive` slightly more often than is strictly necessary.

**Comment by dwrensha** [comment]
I'm guessing that the slowness of the first two examples in test/ring.lean has to do with https://github.com/leanprover/lean4/issues/2055.

**Comment by dwrensha** [comment]
After https://github.com/leanprover-community/mathlib4/pull/2188, the examples in test/ring seem to be going a lot faster.
The commented-out example in test/FieldSimp no longer needs a `maxHeartbeats` adjustment to succeed, but it's still slow enough that I think it makes sense to keep it commented out.

**Comment by fpvandoorn** [comment]
Ok, let's get this merged

bors merge

**Pull request closed without merging**